crates/volta-migrate/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+//! need to be aware that they may be partially applied (if something fails in the process) and should be
+//! able to re-start gracefully from an interrupted migration
+
+#![cfg(feature = "volta-updates")]
+use std::collections::HashSet;
+use std::convert::TryInto;
+use std::fs::{DirEntry, Metadata};
+use std::path::Path;
+
+mod empty;
+mod v0;
+mod v1;
+
+use v0::V0;
+use v1::V1;
+
+use log::debug;
+use volta_core::error::ErrorDetails;
+use volta_core::fs::read_dir_eager;
+use volta_core::layout::volta_home;
+#[cfg(unix)]
+use volta_core::layout::volta_install;
+use volta_core::shim;
+use volta_fail::{Fallible, ResultExt};
+
+enum MigrationState {
+    Empty(empty::Empty),
+    V0(Box<V0>),
+    V1(Box<V1>),
+}
+
+impl MigrationState {
+    fn current() -> Fallible<Self> {
+        /*
+        Triage for determining the current layout version:
+        - Does the V1 layout file exist? If yes then V1
+        - Does Volta Home exist? If yes (Windows) then V0
+            - If yes (Unix) then check if Volta Install is outside shim_dir?
+                - If yes, then V0
+                - If no, then check if $VOLTA_HOME/load.sh exists? If yes then V0
+        - Else Empty
+
+        The extra logic on Unix is necessary because Unix installs can be either inside or outside $VOLTA_HOME
+        If it is inside, then the directory necessarily must exist, so we can't use that as a determination.
+        If it is outside (and for Windows which is always outside), then if $VOLTA_HOME exists, it must be from a
+        previous, V0 installation.
+
+        Going forward, each new version will have an associated layout file, so we can use that as a discriminant
+        */
+
+        let home = volta_home()?;
+        let volta_home = home.root().to_owned();
+
+        if home.layout_file().exists() {
+            return Ok(MigrationState::V1(Box::new(V1::new(volta_home))));
+        }
+
+        if volta_home.exists() {
+            #[cfg(windows)]
+            return Ok(MigrationState::V0(Box::new(V0::new(volta_home))));
+
+            #[cfg(unix)]
+            {
+                let install = volta_install()?;
+                if install.root().starts_with(&volta_home) {
+                    // Installed inside $VOLTA_HOME, so need to look for `load.sh` as a marker
+                    if volta_home.join("load.sh").exists() {
+                        return Ok(MigrationState::V0(Box::new(V0::new(volta_home))));
+                    }
+                } else {
+                    // Installed outside of $VOLTA_HOME, so it must exist from a previous V0 install
+                    return Ok(MigrationState::V0(Box::new(V0::new(volta_home))));
+                }
+            }
+        }
+
+        Ok(MigrationState::Empty(empty::Empty::new(volta_home)))
+    }
+}
+
+pub fn run_migration() -> Fallible<()> {
+    let mut state = MigrationState::current()?;
+
+    loop {
```

Comment:
If I'm reading this right, you don't need a loop at all. You could just have:

```rust
let v1 = match state {
    MigrationState::Empty(e) => e.try_into()?,
    MigrationState::V0(zero) => zero.try_into()?,
    MigrationState::V1(one) => one
};
v1.commit()?;
```

Reply:
We don't strictly need a loop at the moment, that's true. The looping behavior was put in thinking forward to future migrations. My thinking (which I realize wasn't well explained on this point) was that each layout version would only need to have migration code from the _previous_ layout version (and nothing, but that's usually a trivial implementation). Then the migration process would loop through one step at a time, migrating from V0 -> V1 -> V2 -> V3 -> ... -> VX.

This is slightly less efficient when doing the migration, but prevents us from having to write a migration from each previous version to the newest one going forward. That would result in each migration being incrementally more difficult to write than the previous ones.